### ISSUE TITLE
Array arithmetic operators not working as documented

### ISSUE THREAD

**Issue by dvstans** [arangodb/arangodb#7288]
## My Environment

* __ArangoDB Version__:        3.3.19
* __Storage Engine__:             MMFiles
* __Deployment Mode__:       Single Server
* __Deployment Strategy__:   Manual Start
* __Configuration__:               vanilla
* __Infrastructure__:               n/a
* __Operating System__:        Ubuntu 16.04
* __Used Package__:             Debian

## Component, Query & Data

__Affected feature__:
AQL query evaluation

__AQL query (if applicable)__:

Query 1: "for i in doc filter i.x == 0 return i"
Query 2: "for i in doc filter i.x == 1 return i"
Query 3: "for i in doc filter i.x > 1000 return i"

__AQL explain (if applicable)__:
(see below)

__Dataset__:

{_id:doc/1, x:[]}
{_id:doc/2, x:[1]}
{_id:doc/3 x:[1,2,3]}

__Replication Factor & Number of Shards (Cluster only)__:
N/A

__Problem__:
Query 1 returns NO documents
Query 2 returns NO documents
Query 3 returns ALL documents

These results are contrary to current documentation, and don't really make sense (esp. why x > 1000 returns all three documents). It seems any value can be used in query 3 and all documents will always be returned (also true if >- used instead). For arithmetic operators, current documentation (v3.3) states:

"an empty array is converted to 0, an array with one member is converted to the numeric representation of its sole member. Arrays with more members are converted to the number 0."

__Expected result__:

Query 1: doc/1
Query 2: doc/2
Query 3: No documents

**Comment by jsteemann** [comment]
I think the query results are as documented.
The queries do not use any arithmetic operators. Arithmetic operators would be `+`, `-`, `*`, `/` and `%`. 

Instead, the query uses comparison operators (`==` and `>`). These do not convert their operands.
Please check the manual page that lists how values are compared in AQL:
https://docs.arangodb.com/3.3/AQL/Fundamentals/TypeValueOrder.html

There it can be seen that any array compares "bigger" than any numeric value. That is the reason for query 3's result.

**Comment by dvstans** [comment]
I see. I think this is the same unexpected (to me) behavior I ran into previously with NULL being returned as a result in a numeric range comparison. This feels wrong to me from a mathematical point of view as it violates symmetry, but It's probably something that doesn't impact most of your users. In our case we are trying to use AQL to evaluating arbitrary numeric queries on heterogeneous documents containing scientific metadata where we only want results that are mathematically correct (i.e. any comparison between a target and a candidate of different types would evaluate to false). I'm guessing this logic is buried quite deep in the AQL engine and would be hard to modify. Will the up-coming Arango Search allow behavior to be modified?

**Comment by jsteemann** [comment]
Hi @dvstans, you are right, the comparison operator logic is deep inside the internals of AQL.
Changing it will also change the behavior and results of a lot of existing AQL queries, which we obviously like to avoid.
I think a major release (i.e. 4.0) would be a good time to reconsider such fundamentals.
Re arangosearch and its behavior w.r.t. this: I think @gnusi can answer this question from the top of his mind (and I can't), so I'll leave it to him to reply to this.

**Comment by gnusi** [comment]
Hi @dvstans ,
with ArangoSearch you have 2 options:
- treat each element in a an array individually
- "squash" them into a single multi-valued attribute
The second one should give you what you've expected I believe. The behavior is controlled by corresponding property in a view definition, namely 'trackListPositions` which is set to `false` by default.

The following is the link to the recent documentation:
https://docs.arangodb.com/devel/Manual/Views/ArangoSearch/DetailedOverview.html

We also have ArangoSearch tutorial (see `Linking a View with multiple collections (different Link configurations)` section):
https://www.arangodb.com/tutorials/arangosearch/part1-getting-started/

**Comment by Simran-B** [comment]
@dvstans What do you actually want to evaluate with your expressions?

- `i.x == 0`: Do you want to test for an array length of zero, i.e. if the array is empty?
  - `LENGTH(i.x) == 0` is true if x is absent, _null_, _false_, "", [] or {}
  - `IS_ARRAY(i.x) AND LENGTH(i.x) == 0` is true only if it's an empty array
  - `i.x == []` finds documents where x is an array with no or only _null_ elements
- `i.x == 1`: Is this supposed to test for an array length of one or if the value 1 is contained in the array?
  - `LENGTH(i.x) == 1` is true if x is _true_, a number, a string with one character, an array with a single element (including `[null]`) or an object with a single attribute
  - `IS_ARRAY(i.x) AND LENGTH(i.x) == 1` to match documents where x is an array with one element
  - Value 1 among array values: `1 IN i.x` or `i.x ANY == 1`
  - `i.x == [1]` finds documents where x is an array with value 1 as sole element
- `i.x > 1000`: Do you want to test the array length or for a value here?
  - `LENGTH(i.x) > 1000` is true for arrays with over 1k elements, but also strings over 1k characters and objects with more than 1k attributes
  - `IS_ARRAY(i.x) AND LENGTH(i.x) > 1000` matches only arrays with >1000 elements
  - `i.x ANY > 1000` is true if x is an array and at least one of the elements is larger than 1000 (e.g. true for `{ "x": [ null, 2, 3, 5000 ] }`), see [array comparison operators](https://docs.arangodb.com/3.4/AQL/Operators.html#array-comparison-operators)
  - `i.x ALL > 1000` requires x to be an array with all values larger than 1000 but also matches an empty array. Will not match if any of the array elements is _null_.

Something you can do with the Type and Value Order:

    FILTER i.x > [] AND i.x < {}
    FILTER i.x[* RETURN CURRENT > true AND CURRENT < ""] ALL == true

You could achieve the same with functions:

    FILTER IS_ARRAY(i.x) AND LENGTH(i.x) > 0
    FILTER i.x[* RETURN IS_NUMBER(CURRENT)] ALL == true

Both conditions match documents where x is a non-empty array and all its elements are numbers.

**Comment by dvstans** [comment]
Because this is for scientific use (by end-users), what I would prefer is that any comparisons between different types (array-scalar, string-scalar, object-string, etc.) would evaluate to false. In other words, any comparison that doesn't make sense, mathematically, would evaluate to false and thus cull invalid candidate matches.

Because these queries will naturally be complex to begin with, requiring specific functions to test for type/NULL for every comparison will increase the complexity (and learning curve) making it very hard for users to compose/understand their queries. The specific example I provided simply surprised me which is why I investigated it. There is the possibility of a user wanting to treat an array as a vector and compare it's magnitude to a scalar, but I wouldn't expect this from a general search engine (this is not for analytics - just for locating/filtering data).

**Comment by Simran-B** [comment]
I don't think that there's a simple solution. You could do something like `TYPENAME(i.x) == TYPENAME(val)` to ensure both are of the same type, but that's not something you could always put as first filter, because the supported comparison expressions expect and support different combinations of operand types.

If the goal is to hide how AQL works, then you might want to parse the query in e.g. a Foxx service, analyze the AST and add/modify FILTERs as needed. It would be very difficult however. It would be much easier to teach users how AQL works.

Is it correct that you want all three comparisons  `i.x == 0`, `i.x == 1` and `i.x > 1000` to return no documents (none of the three examples you provided) because an array is compared to a number in these cases?

**Comment by dvstans** [comment]
Yes, the desired behavior would be to return no documents for any of the example queries. I am in the process of writing a query pre-processor to address the NULL issue (yes, it's hard!); however, I don't think I can do anything about arrays. You would think it would be easy to teach AQL to scientists, but, the majority of domain scientist at out lab are not particularly programming savvy - and more importantly lack the time/interest needed to learn something like AQL (beyond basic math-like expressions). 

Thanks for the information though!

**Comment by Simran-B** [comment]
Thanks for the confirmation. If you want, you can write up a proposal how it should work in your opinion, also the _null_ issue - a Google Docs document would be good. Please also cover corner-cases including how AQL functions should behave, what operand types they should expect and if functions need to be added/split (I guess that would be necessary for e.g. LENGTH). A statement why you think this should be changed for everyone should also be included.

@jsteemann Is there a ticket / document for things to consider for 4.0 yet?

**Comment by mpoeter** [comment]
Given AQLs close proximity to JavaScript, how about introducing an additional equality operator like JavaScript's [strict equality (===)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Comparison_Operators) instead of changing the semantics of the current equality operator?

**Comment by Simran-B** [comment]
That's an interesting idea. There are no type-checking variants for less/greater than/equal however.

A possible reason might be the confusion for users if both `>=` and `<=` return false for the same value:
https://stackoverflow.com/questions/14533046/why-doesnt-javascript-have-strict-greater-less-than-comparison-operators

The syntax would be uncommon (could possibly confused bitwise operations in some cases) and mathematical users would need to adopt it:
- `>>` strictly greater than
- `>==` / `>>=` strictly greater than or equal
- `<<` strictly less than
- `<==` / `<<=` strictly less than or equal

If we want to reserve `<<` and `>>` for bit shifting, then maybe use `<<<` and `>>>`? (The other two should then be `>>=` and `<<=` I guess.). `>>>` is available in JavaScript for zero-fill right shift bit operations, but it's rarely used I guess.

Would there be any benefit in adding `===` and `!==`? The only use case I can think of would be to distinguish between an explicit null value stored in a document and the absence of an attribute:
- Document: `{ }`
  Filter: `doc.foo == null`
  Result: true (implicit null value)
- Document `{ }`
  Filter: `doc.foo === null`
  Result: false (there is no "instance" of a null value, but really nothing, like _undefined_ in JS)
- Document `{ "foo": null }`
  Filter: `doc.foo == null`
  Result: true (explicit null value)
- Document `{ "foo": null }`
  Filter: `doc.foo === null`
  Result: true (a null value exists as attribute)

In other cases like `LENGTH(doc.attr) == 0`, strict equality operators are of no use, because the function returns a numeric value in all cases and we can't make it type-checking after the fact with a `===`, e.g. to only return true if the attribute is an array with zero elements, but false if the attribute is absent, null, false, "", [] or {}. We would need to change the behavior of the functions and split some functionality into separate functions (e.g. only CHAR_LENGTH to determine the length of a string), or add type-checking variants for a lot of functions (ouch).
